todoist/managers/archive.py

```diff
@@ -0,0 +1,188 @@
+archive = api.items_archive.for_project(project_id)
+
+# Iterate over the list of completed items for the archive
+for item in archive.items():
+    print(item["date_completed"], item["content"])
+```
+"""
+
+from typing import TYPE_CHECKING
+
+from ..models import Item, Model, Section
+from .generic import Manager
+
+if TYPE_CHECKING:
+    from ..api import TodoistAPI
+
+
+class ArchiveManager(Manager):
+
+    object_model = Model
+
+    def __init__(self, api, object_type):
+        # type: (TodoistAPI, str) -> None
+        assert object_type in {"sections", "items"}
+        self.api = api
+        self.cursor = None
+        self.has_more = True
+        self.elements = []
+        self.object_type = object_type
+
+    def reset(self):
+        """Reset internal cache of manager and start iteration from scratch."""
+        self.has_more = True
+        self.cursor = None
+        self.elements[:] = []
+
+    def next_page(self):
+        """Return response for the next page of the archive."""
+        resp = self.api.session.get(
+            self._next_url(),
+            params=self._next_query_params(),
+            headers=self._request_headers(),
+        )
+        resp.raise_for_status()
+        return resp.json()
+
+    def _next_url(self):
+        return "{0}/sync/{1}/archive/{2}".format(
+            self.api.api_endpoint, self.api.api_version, self.object_type
+        )
+
+    def _next_query_params(self):
+        ret = {}
+        if self.cursor:
+            ret["cursor"] = self.cursor
+        return ret
+
+    def _request_headers(self):
+        return {"Authorization": "Bearer {}".format(self.api.token)}
+
+    def _iterate(self):
+        for el in self.elements:
+            yield el
+
+        while True:
+            if not self.has_more:
+                break
+
+            resp = self.next_page()
+            objects = resp[self.object_type]
+
+            self.elements += [self._make_object(data) for data in objects]
+            self.has_more = resp["has_more"]
+            self.cursor = resp.get("next_cursor")
+
+            for obj in objects:
+                yield obj
+
+    def _make_object(self, data):
+        return self.object_model(data, self.api)
+
+
+class SectionsArchiveManagerMaker(object):
+    def __init__(self, api):
+        self.api = api
+
+    def __repr__(self):
+        return "{}()".format(self.__class__.__name__)
+
+    def for_project(self, project_id):
+        """Get manager to iterate over all archived sections for project."""
+        return SectionsArchiveManager(api=self.api, project_id=project_id)
+
+
+class SectionsArchiveManager(ArchiveManager):
+
+    object_model = Section
+
+    def __init__(self, api, project_id):
+        super(SectionsArchiveManager, self).__init__(api, "sections")
+        self.project_id = project_id
+
+    def __repr__(self):
+        return "SectionsArchiveManager(project_id={})".format(self.project_id)
+
+    def sections(self):
```

Comment:
This method is exactly the same as the `items()` in `ItemsArchiveManager` class, maybe we can implement a common generator method for `ArchiveManager` and `sections()` and `items()` are just aliases (if we can't find a good name fitting both items and sections).

Also, maybe instead we can make `ArchiveManager` iteratable itself, with `__iter__()/__next__()` and thus make possible something like:

```
archived_items = api.items_archive.for_project(project_id)
for item in archived_items:
    print(item["date_completed"], item["content"])
```

Reply:
I could be implemented it with just a single line in a class body: `sections = _iterate` (and `items = _iterate`), but I decided that it would probably be better to implement it explicitly to avoid potential confusion of IDEs and code readers. Besides, two functions are not totally equal: they differ in their docstrings and type hints (added type hints just now 😛 )

Reply:
As for providing `__iter__` and `__next__` I would prefer to keep two concerns separate. 

The idea is to have is one manager and many iterators. Clients can make new throwaway iterators as they need with `items()` or `sections()` without worrying about the state of the manager (or without the need to recreate it).

With this saying, I can see that it's not going to work that well in practice because of the shared state, saved in self.cursor, self.has_more and self.elements. I will rewrite this part

Reply:
Rewrote iterators in 32d56f99d8b23a4c8f74668be545c4f6c95c661b making them totally independent from each other.